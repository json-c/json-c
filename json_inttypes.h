
/**
 * @file
 * @brief Do not use, json-c internal, may be changed or removed at any time.
 */
#ifndef _json_inttypes_h_
#define _json_inttypes_h_

#include "json_config.h"

#ifdef JSON_C_HAVE_INTTYPES_H
/* inttypes.h includes stdint.h */
#include <inttypes.h>

#else
#include <stdint.h>

#define PRId64 "I64d"
#define SCNd64 "I64d"
#define PRIu64 "I64u"

#endif

#endif
