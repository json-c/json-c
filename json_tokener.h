/*
 * $Id: json_tokener.h,v 1.10 2006/07/25 03:24:50 mclark Exp $
 *
 * Copyright (c) 2004, 2005 Metaparadigm Pte. Ltd.
 * Michael Clark <michael@metaparadigm.com>
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See COPYING for details.
 *
 */

#ifndef _json_tokener_h_
#define _json_tokener_h_

#include <stddef.h>
#include "json_object.h"

#ifdef __cplusplus
extern "C" {
#endif

enum json_tokener_error {
  json_tokener_success,
  json_tokener_continue,
  json_tokener_error_depth,
  json_tokener_error_parse_eof,
  json_tokener_error_parse_unexpected,
  json_tokener_error_parse_null,
  json_tokener_error_parse_boolean,
  json_tokener_error_parse_number,
  json_tokener_error_parse_array,
  json_tokener_error_parse_object_key_name,
  json_tokener_error_parse_object_key_sep,
  json_tokener_error_parse_object_value_sep,
  json_tokener_error_parse_string,
  json_tokener_error_parse_comment
};

enum json_tokener_state {
  json_tokener_state_eatws,
  json_tokener_state_start,
  json_tokener_state_finish,
  json_tokener_state_null,
  json_tokener_state_comment_start,
  json_tokener_state_comment,
  json_tokener_state_comment_eol,
  json_tokener_state_comment_end,
  json_tokener_state_string,
  json_tokener_state_string_escape,
  json_tokener_state_escape_unicode,
  json_tokener_state_boolean,
  json_tokener_state_number,
  json_tokener_state_array,
  json_tokener_state_array_add,
  json_tokener_state_array_sep,
  json_tokener_state_object_field_start,
  json_tokener_state_object_field,
  json_tokener_state_object_field_end,
  json_tokener_state_object_value,
  json_tokener_state_object_value_add,
  json_tokener_state_object_sep
};

struct json_tokener_srec
{
  enum json_tokener_state state, saved_state;
  struct json_object *obj;
  struct json_object *current;
  char *obj_field_name;
};

#define JSON_TOKENER_MAX_DEPTH 32

struct json_tokener
{
  char *str;
  struct printbuf *pb;
  int depth, is_double, st_pos, char_offset;
  enum json_tokener_error err;
  unsigned int32_t ucs_char;
  char quote_char;
  struct json_tokener_srec stack[JSON_TOKENER_MAX_DEPTH];
};

extern const char* json_tokener_errors[];

extern struct json_tokener* json_tokener_new(void);
extern void json_tokener_free(struct json_tokener *tok);
extern void json_tokener_reset(struct json_tokener *tok);
extern struct json_object* json_tokener_parse(const char *str);
extern struct json_object* json_tokener_parse_verbose(const char *str, enum json_tokener_error *error);
extern struct json_object* json_tokener_parse_ex(struct json_tokener *tok,
						 const char *str, int len);

#ifdef __cplusplus
}
#endif

#endif
