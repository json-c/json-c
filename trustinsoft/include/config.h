
/* Enable RDRAND Hardware RNG Hash Seed */
/* #undef ENABLE_RDRAND */

/* Override json_c_get_random_seed() with custom code */
/* #undef OVERRIDE_GET_RANDOM_SEED */

/* Enable partial threading support */
/* #undef ENABLE_THREADING */

/* Define if .gnu.warning accepts long strings. */
/* #undef HAS_GNU_WARNING_LONG */

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H

/* Define to 1 if you have the <endian.h> header file. */
#define HAVE_ENDIAN_H

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H

/* Define to 1 if you have the <limits.h> header file. */
#define HAVE_LIMITS_H

/* Define to 1 if you have the <locale.h> header file. */
#define HAVE_LOCALE_H

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H

/* Define to 1 if you have the <stdarg.h> header file. */
#define HAVE_STDARG_H

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H

/* Define to 1 if you have the <syslog.h> header file. */
#define HAVE_SYSLOG_H 1

/* Define to 1 if you have the <sys/cdefs.h> header file. */
#define HAVE_SYS_CDEFS_H

/* Define to 1 if you have the <sys/param.h> header file. */
#define HAVE_SYS_PARAM_H 1

/* Define to 1 if you have the <sys/random.h> header file. */
#define HAVE_SYS_RANDOM_H

/* Define to 1 if you have the <sys/resource.h> header file. */
#define HAVE_SYS_RESOURCE_H

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the <xlocale.h> header file. */

/* Define to 1 if you don't have `vprintf' but do have `_doprnt.' */
/* #undef HAVE_DOPRNT */

/* Has atomic builtins */
#define HAVE_ATOMIC_BUILTINS

/* Define to 1 if you have the declaration of `INFINITY', and to 0 if you
   don't. */
#define HAVE_DECL_INFINITY

/* Define to 1 if you have the declaration of `isinf', and to 0 if you don't.
   */
#define HAVE_DECL_ISINF

/* Define to 1 if you have the declaration of `isnan', and to 0 if you don't.
   */
#define HAVE_DECL_ISNAN

/* Define to 1 if you have the declaration of `nan', and to 0 if you don't. */
#define HAVE_DECL_NAN

/* Define to 1 if you have the declaration of `_finite', and to 0 if you
   don't. */
/* #undef HAVE_DECL__FINITE */

/* Define to 1 if you have the declaration of `_isnan', and to 0 if you don't.
   */
/* #undef HAVE_DECL__ISNAN */

/* Define to 1 if you have the `open' function. */
#define HAVE_OPEN

/* Define to 1 if you have the `realloc' function. */
#define HAVE_REALLOC

/* Define to 1 if you have the `setlocale' function. */
#define HAVE_SETLOCALE

/* Define to 1 if you have the `snprintf' function. */
#define HAVE_SNPRINTF


/* Define to 1 if you have the `strcasecmp' function. */
#define HAVE_STRCASECMP 1

/* Define to 1 if you have the `strdup' function. */
#define HAVE_STRDUP

/* Define to 1 if you have the `strerror' function. */
#define HAVE_STRERROR

/* Define to 1 if you have the `strncasecmp' function. */
#define HAVE_STRNCASECMP 1

/* Define to 1 if you have the `uselocale' function. */

/* Define to 1 if you have the `vasprintf' function. */
#define HAVE_VASPRINTF

/* Define to 1 if you have the `vprintf' function. */
#define HAVE_VPRINTF

/* Define to 1 if you have the `vsnprintf' function. */
#define HAVE_VSNPRINTF

/* Define to 1 if you have the `vsyslog' function. */
#define HAVE_VSYSLOG 1

/* Define if you have the `getrandom' function. */
#define HAVE_GETRANDOM

/* Define if you have the `getrusage' function. */
#define HAVE_GETRUSAGE

#define HAVE_STRTOLL
#if !defined(HAVE_STRTOLL)
#define strtoll strtoll
/* #define json_c_strtoll strtoll*/
#endif

#define HAVE_STRTOULL
#if !defined(HAVE_STRTOULL)
#define strtoull strtoull
/* #define json_c_strtoull strtoull */
#endif

/* Have __thread */
#define HAVE___THREAD

/* Public define for json_inttypes.h */
#define JSON_C_HAVE_INTTYPES_H 1

/* Name of package */
#define PACKAGE "json-c"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "json-c@googlegroups.com"

/* Define to the full name of this package. */
#define PACKAGE_NAME "json-c"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "json-c 0.15.99"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "json-c"

/* Define to the home page for this package. */
#define PACKAGE_URL "https://github.com/json-c/json-c"

/* Define to the version of this package. */
#define PACKAGE_VERSION "0.15.99"

/* The number of bytes in type int */
#define SIZEOF_INT 4

/* The number of bytes in type int64_t */
#define SIZEOF_INT64_T 8

/* The number of bytes in type long */
#define SIZEOF_LONG 8

/* The number of bytes in type long long */
#define SIZEOF_LONG_LONG 8

/* The number of bytes in type size_t */
#define SIZEOF_SIZE_T 8

/* The number of bytes in type ssize_t */
#define SIZEOF_SSIZE_T 8

/* Specifier for __thread */
#define SPEC___THREAD __thread

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS

/* Version number of package */
#define VERSION "0.15.99"

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* Define to `unsigned int' if <sys/types.h> does not define. */
/* #undef size_t */
