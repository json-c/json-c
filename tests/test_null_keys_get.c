/*
 * Tests if binary strings are supported.
 */

#include "config.h"
#include <stdio.h>
#include <string.h>

#include "json_inttypes.h"
#include "json_object.h"
#include "json_tokener.h"

int main(void)
{
	/* this test has embedded null characters in the key and value.
	 * check that it's still included after deserializing. */
	const char *input = "{ \"foo\\u0000bar\": \"qwerty\\u0000asdf\" }";
	const char *expected_key = "foo\0bar";
	const int expected_key_len = 7;
	const char *expected_key_printable = "foo\\0bar";
	const char *expected_value = "qwerty\0asdf";
	const int expected_value_len = 11;
	const char *expected_value_printable = "qwerty\\0asdf";

	struct json_object *parsed = json_tokener_parse(input);

	printf("Parsed input: %s\n", input);
	printf("Result is ");
	if (parsed == NULL)
	{
		printf("NULL (error!)\n");
		return 1; // Error somewhere
	}
	else if (!json_object_is_type(parsed, json_type_object))
	{
		printf("not `json_type_object` (error!)\n");
		return 1; // Error somewhere
	}
	else
	{
		printf("`json_type_object`\n");
	}

	// Check nothing odd happened in parsing
	if (json_object_object_length(parsed) != 1)
	{
		printf("Should contain only one field (has %d) (error!)",
		       json_object_object_length(parsed));
		return 1;
	}
	json_bool key_present = json_object_object_get_ex(parsed, expected_key, NULL);
	if (key_present)
	{
		printf("The key \"%s\" should not be present when calling "
		       "`json_object_object_get_ex` "
		       "(the real key contains a NUL character). (error!)\n",
		       expected_key);
		return 1;
	}

	// Check the key is present
	struct json_object *string;
	key_present = json_object_object_get_ex_len(parsed, expected_key, 7, &string);
	if (!key_present)
	{
		printf("The key \"%s\" should be present when calling "
		       "`json_object_object_get_ex_len` (error!)\n",
		       expected_key_printable);
		return 1;
	}
	else if (string == NULL)
	{
		printf("The key \"%s\" was present when calling "
		       "`json_object_object_get_ex_len`, but got NULL (error!)\n",
		       expected_key_printable);
		return 1;
	}
	struct json_object *from_get_len = json_object_object_get_len(parsed, expected_key, 7);
	if (string != from_get_len)
	{
		printf("The value returned from `json_object_object_get_len` should be the "
		       "same as the one from `json_object_object_get_ex_len` (error!)\n"
		       "Got %p from `json_object_object_get_ex_len` but %p from "
		       "`json_object_object_get_len`\n",
		       string, from_get_len);
		return 1;
	}
	else
	{
		printf("Key was present and same for "
		       "`json_object_object_get_ex_len` and `json_object_object_get_len`\n");
	}

	// Check the value is right
	if (!json_object_is_type(string, json_type_string))
	{
		printf(
		    "Value is wrong type, expected `json_type_string` (%d) but got %d (error!)\n",
		    json_type_string, json_object_get_type(string));
		return 1;
	}
	else
	{
		printf("Value is right type, `json_type_string`\n");
	}
	const int actual_value_len = json_object_get_string_len(string);
	const char *actual_value = json_object_get_string(string);
	if (actual_value_len != expected_value_len)
	{
		printf("Value is wrong length, expected %d but got %d (error!)\n",
		       expected_value_len, actual_value_len);
		return 1;
	}
	else if (memcmp(expected_value, actual_value, actual_value_len) != 0)
	{
		printf("Expected \"%s\" but got \"", expected_value_printable);
		for (int i = 0; i < actual_value_len; ++i)
		{
			if (actual_value[i] == '\0')
			{
				puts("\\0");
			}
			else
			{
				putchar(actual_value[i]);
			}
		}
		printf("\"\n");
		return 1;
	}
	else
	{
		printf("Expected value matches actual\n");
	}

	json_object_put(parsed);
	printf("PASS\n");
	return 0;
}
