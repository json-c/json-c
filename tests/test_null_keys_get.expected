Parsed input: { "foo\u0000bar": "qwerty\u0000asdf" }
Result is `json_type_object`
Key was present and same for `json_object_object_get_ex_len` and `json_object_object_get_len`
Value is right type, `json_type_string`
Expected value matches actual
PASS
